### ISSUE TITLE
Allow prepending a custom user agent

### ISSUE THREAD

**Issue by everett-toews** [openstacknetsdk/openstack.net#478]
Give your users the ability to _prepend_ a custom user agent. So if the existing user agent is

    openstack.net/1.1.0

Provide a method for your users to easily prepend their own user agent so they can further customize it.

    users-app/0.1.2 openstack.net/1.1.0 

This will give us much better insight into how people are using openstack.net. This needs to be well documented and part of __any__ examples that are written for openstack.net so people just copy/paste it in.

**Comment by sharwell** [comment]
Hi @everett-toews,

Welcome to the .NET SDK project! :sparkles: 

Thanks for taking time to write up this request. Special consideration for application-specific user agents is a topic that hasn't been discussed previously, but it makes complete sense that a both a user and a service provider would value this functionality.

The .NET SDK allows complete customization of the user agent, including but not limited to prepending a string. Note that this does *not* mean it's particularly easy. Since the V1 and V2 use completely different approaches to sending HTTP requests, each SDK will need to address this in a suitable manner.

### SDK V1

#### Current implementation

Each provider class in the V1 SDK includes a protected method [`BuildDefaultRequestSettings`](http://openstacknetsdk.org/docs/html/M_net_openstack_Providers_Rackspace_ProviderBase_1_BuildDefaultRequestSettings.htm). This method can be overridden to specify the user agent value used by that provider instance. The following code shows how this method may be used to prepend a value to the default user agent used by the library.

```csharp
protected override RequestSettings BuildDefaultRequestSettings(IEnumerable<HttpStatusCode> non200SuccessCodes = null)
{
  RequestSettings settings = base.BuildDefaultRequestSettings(non200SuccessCodes);
  settings.UserAgent = "users-app/0.1.2 " + settings.UserAgent;
  return settings;
}
```

#### Potential improvements

A new `ApplicationUserAgent` property could be added to the `ProviderBase<TProvider>` class. Since the V1 SDK represents user agents as a simple string value, the defalut implementation of `BuildDefaultRequestSettings` would need to be updated to prepend this value to the default value whenever it is specified. No validation would be provided to ensure that the resulting header conforms to the form described in RFC7231.

### SDK V2

#### Current implementation

The V2 SDK allows substantially more customization, and there are multiple ways the user agent header could be modified.

1. The application can extend an existing service client class and override the `ServiceClient.PrepareRequestImpl<T>` method. This method can update, replace, or remove the value of the [`HttpRequestHeaders.UserAgent`](https://msdn.microsoft.com/en-us/library/system.net.http.headers.httprequestheaders.useragent.aspx) property.
2. The application can attach a listener to the `ServiceClient.BeforeAsyncWebRequest` event, and modify the request message (including the user agent).
3. The application can provide a custom `HttpClient` and/or `HttpMessageHandler` implementation for use with the service clients.

Each of these strategies has advantages and disadvantages, but none of them would be considered an ideal solution for the problem as presented.

#### Potential improvements

The easiest way to expose this functionality throughout the V2 SDK is creating a new instance property `ServiceClient.ApplicationUserAgent` with the type [`ProductInfoHeaderValue`](https://msdn.microsoft.com/en-us/library/system.net.http.headers.productinfoheadervalue.aspx). If this value is not `null`, it is added to the list of user agent values before the standard header for the SDK is added.

The `HttpClient` library used by the V2 SDK performs validation of the user agent header according to RFC2616, which would also be valid in RFC7231.

### Documentation concerns

Updating the documentation in the described manner might not be feasible. Many examples, such as the one for [`IObjectStorageProvider.CreateObject`](http://openstacknetsdk.org/docs/html/M_net_openstack_Core_Providers_IObjectStorageProvider_CreateObject.htm), are written in a form that assumes the provider instance is initialized elsewhere. The updates to demonstrate application-specific user agents would likely be focused on examples showing the creation of clients. At the current time, no such examples exist in the generated documentation for the SDK (at least in the form of code that can be copied and pasted). Examples of this form do appear in other locations, such as [developer.rackspace.com](https://developer.rackspace.com/docs/cloud-files/getting-started/?lang=dot-net).

**Comment by carolynvs** [comment]
Fixed by #482 which was released in v1.3.6.0.

**Comment by everett-toews** [comment]
:+1:
